homeassistant/components/numato/__init__.py

```diff
@@ -0,0 +1,211 @@
+    if rng[0] > rng[1]:
+        raise vol.Invalid(f"Lower range bound must come first: {rng}")
+    return rng
+
+
+def float_range(rng):
+    """Validate the input array to describe a range by two floats."""
+    try:
+        coe = vol.Coerce(float)
+        coe(rng[0])
+        coe(rng[1])
+    except vol.CoerceInvalid:
+        raise vol.Invalid(f"Only int or float values are allowed: {rng}")
+    if len(rng) != 2:
+        raise vol.Invalid(f"Only two numbers allowed in a range: {rng}")
+    if rng[0] > rng[1]:
+        raise vol.Invalid(f"Lower range bound must come first: {rng}")
+    return rng
+
+
+def adc_port_number(num):
+    """Validate input number to be in the range of ADC enabled ports."""
+    try:
+        num = int(num)
+    except (ValueError):
+        raise vol.Invalid(f"Port numbers must be integers: {num}")
+    if num not in range(1, 8):
+        raise vol.Invalid(f"Only port numbers from 1 to 7 are ADC capable: {num}")
+    return num
+
+
+ADC_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Optional(CONF_SRC_RANGE, default=DEFAULT_SRC_RANGE): int_range,
+        vol.Optional(CONF_DST_RANGE, default=DEFAULT_DST_RANGE): float_range,
+        vol.Optional(CONF_DST_UNIT, default=DEFAULT_UNIT): cv.string,
+    }
+)
+
+PORTS_SCHEMA = vol.Schema({cv.positive_int: cv.string})
+
+IO_PORTS_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_PORTS): PORTS_SCHEMA,
+        vol.Optional(CONF_INVERT_LOGIC, default=DEFAULT_INVERT_LOGIC): cv.boolean,
+    }
+)
+
+DEVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ID): cv.positive_int,
+        CONF_BINARY_SENSORS: IO_PORTS_SCHEMA,
+        CONF_SWITCHES: IO_PORTS_SCHEMA,
+        CONF_SENSORS: {CONF_PORTS: {adc_port_number: ADC_SCHEMA}},
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: {
+            CONF_DEVICES: vol.All(cv.ensure_list, [DEVICE_SCHEMA]),
+            vol.Optional(CONF_DISCOVER, default=DEFAULT_DEV): vol.All(
+                cv.ensure_list, [cv.string]
+            ),
+        },
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+def setup(hass, config):
+    """Initialize the integration by discovering available Numato devices."""
+    hass.data[DOMAIN] = config[DOMAIN][CONF_DEVICES]
+
+    gpio.discover(config[DOMAIN][CONF_DISCOVER])
+
+    _LOGGER.info(
+        "Initializing Numato 32 port USB GPIO expanders with IDs: %s",
+        ", ".join(str(d) for d in gpio.devices),
+    )
+
+    def cleanup_gpio(event):
+        """Stuff to do before stopping."""
+        _LOGGER.debug("Clean up Numato GPIO")
+        gpio.cleanup()
+        PORTS_IN_USE.clear()
+
+    def prepare_gpio(event):
+        """Stuff to do when home assistant starts."""
+        _LOGGER.debug("Setup cleanup at stop for Numato GPIO")
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, cleanup_gpio)
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, prepare_gpio)
+
+    return True
+
+
+PORTS_IN_USE = dict()
```

Comment:
Store the ports in hass.data instead.

Reply:
Since I don't have a hass object in these functions and don't want to store it on module level, I'll change the API to be a class for each device created in setup() after discovery and pass the API objects via hass.data[DOMAIN]. Just like in octoprint. Good style? Any objections or suggestions?

Reply:
Yeah, we can wrap the gpio interface in a class and store the instance of that in hass.data. Then give each entity in each platform a reference to that instance.

Reply:
Done.